File: app/src/main/java/com/nirwashh/rickandmortyapp/characters/presentation/detail/viewmodel/CharacterDetailViewModelFactory.java
package com.nirwashh.rickandmortyapp.characters.presentation.detail.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.nirwashh.rickandmortyapp.episodes.domain.EpisodesInteractor;
import com.nirwashh.rickandmortyapp.episodes.presentation.mapper.EpisodeDomainToUi;
import com.nirwashh.rickandmortyapp.locations.domain.LocationInteractor;

public class CharacterDetailViewModelFactory implements ViewModelProvider.Factory {
    EpisodesInteractor episodesInteractor;
    LocationInteractor locationInteractor;
    EpisodeDomainToUi episodeDomainToUi;

    public CharacterDetailViewModelFactory(EpisodesInteractor episodesInteractor, LocationInteractor locationInteractor, EpisodeDomainToUi episodeDomainToUi) {
        this.episodesInteractor = episodesInteractor;
        this.locationInteractor = locationInteractor;
        this.episodeDomainToUi = episodeDomainToUi;
    }

    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new CharacterDetailViewModel(episodesInteractor, locationInteractor, episodeDomainToUi);
    }
}

File: app/src/main/java/com/nirwashh/rickandmortyapp/characters/presentation/detail/viewmodel/CharacterDetailViewModelFactory.kt
package com.nirwashh.rickandmortyapp.characters.presentation.detail.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.nirwashh.rickandmortyapp.episodes.domain.EpisodesInteractor
import com.nirwashh.rickandmortyapp.locations.domain.LocationInteractor

class CharacterDetailViewModelFactory(
    private val episodesInteractor: EpisodesInteractor,
    private val locationInteractor: LocationInteractor
) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return CharacterDetailViewModel(episodesInteractor, locationInteractor) as T
    }
}